File: Programmers/Lv. 3/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    # n개의 노드, 간선 정보 edge
    lines = [[] for i in range(n+1)]
    for a, b in edge:
        lines[a].append(b)
        lines[b].append(a)
    
    distances = [-1] * (n+1)
    distances[1] = 0
    
    q = deque([1])
    while q:
        now = q.popleft()
        for node in lines[now]:
            if distances[node] == -1:
                print(node)
                distances[node] = distances[now] + 1
                q.append(node)
    
    return distances.count(max(distances))

File: Programmers/Lv. 3/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>

using namespace std;

int solution(int n, vector<vector<int>> edge) {
    vector<int> distances;
    vector<bool> visited;
    map<int, vector<int>> maps;
    for (int i = 0; i < n+1; i++)
    {
        distances.push_back(0);
        
        visited.push_back(false);
        
        maps[i] = {};
    }    
    visited[1] = true;
    
    for (vector<int> e : edge)
    {
        maps[e[0]].push_back(e[1]);
        maps[e[1]].push_back(e[0]);
    }
    
    queue<int> q;
    q.push(1);
    
    while (!q.empty())
    {
        int n = q.front();
        q.pop();
        
        for (int i : maps[n])
        {
            if (!visited[i])
            {
                visited[i] = true;
                
                distances[i] = distances[n] + 1;
                
                q.push(i);
            }
        }    
    }
    
    int m = *max_element(distances.begin(), distances.end());
    
    return count(distances.begin(), distances.end(), m);
}